webapp/api/launchpad.py

```diff
@@ -1,6 +1,105 @@
-from launchpadlib.launchpad import Launchpad
+from hashlib import md5
+from webapp.api.requests import Session
 
 
-launchpad = Launchpad.login_anonymously(
-    "storefront", "production", version="devel"
-)
+class Launchpad:
+    """
+    Provides authentication for Launchpad.
+    """
+
+    BASE_URL = "https://api.launchpad.net"
+
+    def __init__(self, username, token, signature, session=Session()):
+        self.username = username
+        self.session = session
+        self.session.headers["Accept"] = "application/json"
+        self.session.headers["Authorization"] = (
+            f'OAuth oauth_version="1.0", '
+            f'oauth_signature_method="PLAINTEXT", '
+            f"oauth_consumer_key={username}, "
+            f'oauth_token="{token}", '
+            f'oauth_signature="&{signature}"'
+        )
+
+    def _request(self, method, url, params={}, data={}):
+        """
+        Makes a raw HTTP request and returns the response.
+        """
+        if not url.startswith("http"):
+            url = f"{self.BASE_URL}/devel/{url}"
+
+        response = self.session.request(method, url, params=params, data=data)
+        response.raise_for_status()
+
+        return response
+
+    def get(self, resource, params=None):
+        """
+        Makes a raw HTTP GET request and returns the response.
+        """
+        response = self._request("GET", resource, params=params)
+
+        return response.json()
+
+    def post(self, resource, data=None):
+        """
+        Makes a raw HTTP POST request and returns the response.
+        """
+        response = self._request("POST", resource, data=data)
+
+        return response.json() if response.content else None
+
+    def get_collection_entries(self, resource, params=None):
+        """
+        Return collection items from the API
+        """
+        collection = self.get(resource, params)
+
+        return collection.get("entries", [])
+
+    def get_snap_by_store_name(self, snap_name):
+        """
+        Return an Snap from the Launchpad API by store_name
+        """
+        snaps = self.get_collection_entries(
+            "+snaps",
+            {
+                "ws.op": "findByStoreName",
+                "owner": "/~build.snapcraft.io",
+                "store_name": snap_name,
+            },
+        )
+
+        # The Launchpad API only allow to perform a find by store_name
+        # but we are only interested in exactly this one
+        if snaps and snaps[0]["store_name"] == snap_name:
+            return snaps[0]
+
+        return None
+
+    def new_snap(self, snap_name, git_url):
+        """
+        Create an ISnap in Launchpad
+        """
+        data = {
+            "ws.op": "new",
+            "owner": f"/~{self.username}",
```

Comment:
Is the owner necessarily `self.username`? Mightn't the owner be different from the current user?

Reply:
All the snaps in launchpad are created by the same user. In this case (and always unless trying with the staging username) self.username is `os.getenv("LP_API_USERNAME")`